libindy/src/api/crypto.rs

```diff
@@ -557,4 +561,190 @@ pub  extern fn indy_crypto_anon_decrypt(command_handle: IndyHandle,
+///        "alg": "authcrypt",
+///        "recipients": [
+///            {
+///                "encrypted_key": anoncrypt(encrypted_cek|sender_vk|nonce)
+///                "header": {
+///                    "kid": "b64URLencode(ver_key)"
+///                }
+///            },
+///        ],
+///    })"
+///    "iv": <b64URLencode()>,
+///    "ciphertext": <b64URLencode(encrypt({'@type'...}, cek)>,
+///    "tag": <b64URLencode()>
+/// }
+///
+/// Alternative example in using anoncrypt alg is defined below:
+/// {
+///    "protected": "b64URLencode({
+///        "enc": "xsalsa20poly1305",
+///        "typ": "JWM/1.0",
+///        "alg": "anoncrypt",
+///        "recipients": [
+///            {
+///                "encrypted_key": <b64URLencode(anoncrypt(cek))>,
+///                "header": {
+///                    "kid": "b64URLencode(ver_key)"
+///                }
+///            },
+///        ],
+///    })"
+///    "iv": <b64URLencode(iv)>,
+///    "ciphertext": <b64URLencode(encrypt({'@type'...}, cek)>,
+///    "tag": <b64URLencode(authentication tag)>
+/// }
+///
+///
+/// #Errors
+/// Common*
+/// Wallet*
+/// Ledger*
+/// Crypto*
+#[no_mangle]
+pub extern fn indy_pack_message(
+    command_handle: IndyHandle,
+    wallet_handle: IndyHandle,
+    message: *const u8,
+    message_len: u32,
+    receiver_keys: *const c_char,
+    sender: *const c_char,
+    cb: Option<extern "C" fn(xcommand_handle: i32, err: ErrorCode, jwe_data: *const u8, jwe_len: u32)>,
+) -> ErrorCode {
+    trace!("indy_pack_message: >>> wallet_handle: {:?}, message: {:?}, message_len {:?},\
+            receiver_keys: {:?}, sender: {:?}", wallet_handle, message, message_len, receiver_keys, sender);
+
+    check_useful_c_byte_array!(message, message_len, ErrorCode::CommonInvalidParam2, ErrorCode::CommonInvalidParam3);
+    check_useful_c_str!(receiver_keys, ErrorCode::CommonInvalidParam4);
+    check_useful_c_str_empty_accepted!(sender, ErrorCode::CommonInvalidParam5);
+    check_useful_c_callback!(cb, ErrorCode::CommonInvalidParam6);
+
+    trace!("indy_pack_message: entities >>> wallet_handle: {:?}, message: {:?}, message_len {:?},\
+            receiver_keys: {:?}, sender: {:?}", wallet_handle, message, message_len, receiver_keys, sender);
+
+    let result = CommandExecutor::instance().send(Command::Crypto(CryptoCommand::PackMessage(
+        message,
+        receiver_keys,
+        sender,
+        wallet_handle,
+        Box::new(move |result| {
+            let (err, jwe) = result_to_err_code_1!(result, Vec::new());
+            trace!("indy_auth_pack_message: jwe: {:?}", jwe);
+            let (jwe_data, jwe_len) = ctypes::vec_to_pointer(&jwe);
+            cb(command_handle, err, jwe_data, jwe_len)
+        }),
+    )));
+
+    let res = result_to_err_code!(result);
+
+    trace!("indy_auth_pack_message: <<< res: {:?}", res);
+
+    res
+}
+
+
+/// Unpacks a message packed using indy_pack_message which follows the wire message format
```

Comment:
I've added experimental to this so that consumers know that it's experimental. From our discussion earlier though because this is only using keys and key referencing is being handled at a different layer, I don't think this will cause problems.